pex/cli/commands/lockfile/lockfile.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, print_function
+
+import os
+
+from pex.compatibility import urlparse
+from pex.distribution_target import DistributionTarget
+from pex.resolve.locked_resolve import LockedResolve
+from pex.resolve.resolver_configuration import ResolverVersion
+from pex.sorted_tuple import SortedTuple
+from pex.third_party.packaging import tags
+from pex.third_party.pkg_resources import Requirement, RequirementParseError
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import attr  # vendor:skip
+    from typing import (
+        Iterable,
+        Iterator,
+        List,
+        Mapping,
+        Optional,
+        Tuple,
+    )
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class _RankedLock(object):
```

Comment:
Does the ranking described in #1414 answer that, perhaps?

Reply:
Agreed. Warm up for adding the doc below in the form of the answers.

> From reading, it seems like you are generating multiple conceptual resolves which are each independently usable, and then selecting from them for a particular consumer.

Locks aside - this is fundamentally what a PEX is. It contains multiple resolves and selects the correct one at runtime. A lock file for a Pex then has to do the same thing.

> Does that mean there is never an exact match for the consumer?

No, although it's likely you don't have the correct notion of what exact match means. For a lock generated for --platform you can have an exact match if you export for the same --platform. And that's about it. Otherwise, the lock was generated for an interpreter and the only way to be sure the match is exact in that case is if the interpreter is the exact same interpreter, which means the exact same ordered set of supported tags and the exact same environment marker environment dict.

That aside and accepting the vague notion of exact match, if you did only support exact matches, then you could never support LockStyle.SOURCES or the forthcoming LockStyle.CROSS_PLATFORM. Even those aside, with a LockStyle.STRICT for a set of requirements that encloses a - say - universal wheel set - you'd not want to fail an export just because the target being used on this machine doesn't match the one on the machine used to generate the lock in the 1st place.